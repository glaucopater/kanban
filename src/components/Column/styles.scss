@import "../../styles/sass/variables.scss";
@import "../../styles/sass/mixins.scss";

.Column {
  border: 1px solid $white;
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  flex-basis: auto;
  min-width: 33%;
  color: white;
  text-align: center;
  > h3 {
    padding: $offsetWidth;
    margin: $offsetWidth;
  }
  > header  {
    display: flex;
    flex-direction: column;
    border-bottom: 2px solid $white;
    place-content: center;
    place-items: center;
    padding: 4px;
  }
}


.ToDo {
  background: $color-column-todo;
}

.ToDo > header {
  background: $color-column-header-todo;
}

.InProgress {
  background: $color-column-inprogress;
}

.InProgress header {
  background: $color-column-header-inprogress;
}

.Done {
  background: $color-column-done;
}

.Done header {
  background: $color-column-header-done;
}

.add-button {
  @include button();
}
