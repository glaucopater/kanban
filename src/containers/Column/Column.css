.Column {
  border: 1px solid black; 
  display: flex;
  flex-direction: column; 
  flex-grow: 1;
  flex-basis: auto;
  color: white;
}

.ToDo {
  background: #b5e4f8;
}

div.Column.ToDo > header {
  background: #0093e1;
}
 
.InProgress {
  background: #fdc0c2;
}

.InProgress header {
  background: #f50057;
}

.Done {
  background: #bac2cb;
}

.Done header {
  background: #0b2542;
}


.add-button {  
  background: transparent; 
  font-size: 1.2em;
  float: right;
  outline: none;
  color: white;
  border: none;
}