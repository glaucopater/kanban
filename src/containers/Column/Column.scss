@import "../../sass/variables.scss";
@import "../../sass/mixins.scss";

.Column {
  border: 1px solid black; 
  display: flex;
  flex-direction: column; 
  flex-grow: 1;
  flex-basis: auto;
  color: white;
  text-align: center;
}

.ToDo {
  background: $color-column-todo;
}

div.Column.ToDo > header {
  background: $color-column-header-todo;
}

.InProgress {
  background: $color-column-inprogress;
}

.InProgress header {
  background: $color-column-header-inprogress;
}

.Done {
  background: $color-column-done;
}

.Done header {
  background: $color-column-header-done;
}

.add-button {
  @include button();
}