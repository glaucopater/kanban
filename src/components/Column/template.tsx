import React from "react";
import { Card } from "../Card/";
import "./styles.scss";

interface IColumnProps {
  category: string;
  category_tasks: any;
  createCard: any;
  updateCard: (card: any, destinationCategory: string) => void;
  moveCard: (cardId: number, sourceCategory: string, destinationCategory: string) => void;
  removeCardFromColumn: (id: number, category: string) => void;
}

interface IColumnState {
  category: string;
  category_tasks: string[];
}

export const Column: React.FC<IColumnProps> = (props) => {
  const [state, setState] = React.useState<IColumnState>(props);
  React.useEffect(() => {
    setState({
      category: props.category,
      category_tasks: [],
    });
  }, [props.category]);

  const handleAdd = () => {
    const maxIdPerCategory = Math.max(
      ...props.category_tasks.map((item: any) => item.id),
      0
    );
    const nextId = maxIdPerCategory + 1;
    const new_task = { id: nextId, text: "Card " + nextId };
    const category_tasks = [new_task, ...props.category_tasks];
    setState({ ...state, category: state.category, category_tasks });
    props.createCard(new_task, props.category);
  };

  const handleRemove = (card_id_to_remove: number) => {
    const category_tasks = props.category_tasks.filter((task: any) => {
      return task.id !== card_id_to_remove;
    });
    setState({ ...state, category_tasks });
    props.removeCardFromColumn(card_id_to_remove, state.category);
  };

  const handleUpdate = (card: any) => {
    const new_task = card;
    const category_tasks = [new_task, ...props.category_tasks];
    setState({ ...state, category_tasks });
    props.updateCard(new_task, props.category);
  };

  const handleOnDragOver = (event: any) => {
    event.preventDefault();
  };

  const handleOnDrop = (event: any) => {
    let id = event.dataTransfer.getData("id");
    let sourceCategory = event.dataTransfer.getData("sourceCategory");
    moveTask(id, sourceCategory);
  };

  const moveTask = (taskId: number, sourceCategory: string) => {
    if (sourceCategory !== state.category) {
      props.moveCard(taskId, sourceCategory, state.category);
    }
  };

  const category_tasks = props.category_tasks;
  const category = props.category ? props.category : "";
  const cards = category_tasks.map((task: any, key: number) => {
    return (
      <Card
        key={key}
        id={task.id}
        text={task.text}
        category={category}
        removeCard={handleRemove}
        updateCard={handleUpdate}
      />
    );
  });

  return (
    <div
      className={"Column " + category}
      onDrop={handleOnDrop}
      onDragOver={handleOnDragOver}
    >
      <header>
        <h3>{category.split(/(?=[A-Z])/).join(" ")}</h3>
        <button className="add-button" onClick={handleAdd}>
          +
        </button>
        <div className="counter">({category_tasks.length})</div>
      </header>
      <div className="task-container droppable">{cards}</div>
    </div>
  );
};

