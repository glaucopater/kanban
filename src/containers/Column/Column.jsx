import React from "react";  
import Card from "../../components/Card/Card"; 
import "./Column.css";
import { DATA } from "../../common/constants";
import  "../../common/helpers";
import { getRandomInt } from "../../common/helpers";


export default class Column extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      category: this.props.category,
      category_tasks: []
    };
    this.handleRemove = this.handleRemove.bind(this);
    this.handleAdd = this.handleAdd.bind(this);
 
  }
 

  handleOnClick(){
    console.log("click");
  }

  handleAdd = () => {
    const randomIndex = getRandomInt(DATA.length);
    console.log("click add item ",0,DATA.length, randomIndex);
    const random_task = DATA.filter((task) => { return task.index === randomIndex && task.category === this.state.category } );

    const category_tasks = [...random_task, ...this.state.category_tasks];
    this.setState({category_tasks });
  }

  handleRemove = (index_to_remove) => {
    console.log("click remove item ", index_to_remove);
    const category_tasks = this.state.category_tasks.filter((task) => { return  task.index !== index_to_remove } );
    this.setState({category_tasks});
  }

  componentDidMount() {
    const category_tasks = DATA.filter((task, key) => { return task.category === this.state.category })
    this.setState({category_tasks});
  }


  render() {
    
    const cards = this.state.category_tasks.map((task, key) => { 
          return <Card key={key} 
          index={task.index} 
          text={task.text}
          bgcolor={task.bgcolor} 
          category={this.state.category} 
          handleRemove={this.handleRemove}/>
        });
    
    const category = this.props.category ? this.props.category : "";
    console.log(category);
    return (
      <div className={"Column " + category }>
      <header>
        <h1>{category}</h1>
        <h2>({this.state.category_tasks.length})</h2>
        <button onClick={this.handleAdd}>+</button>
      </header>  
        <div className="task-container"> 
          {cards}
        </div>
      </div>
    );
  }
}
