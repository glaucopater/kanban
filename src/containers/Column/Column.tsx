import React from "react";
import Card from "../../components/Card/Card";
import "./Column.scss";


interface IColumnProps {
  category: string;
  category_tasks: any;
  createCard: any;
  updateCard: (card: any, destinationCategory: string) => void;
  moveCard: (cardId: number, sourceCategory: string, destinationCategory: string) => void;
  removeCardFromColumn: (id: number, category: string) => void;
}

interface IColumnState {
  category: string;
  category_tasks: string[];
}

export default class Column extends React.Component<IColumnProps, IColumnState> {
  constructor(props: IColumnProps) {
    super(props);
    this.state = {
      category: this.props.category,
      category_tasks: [],
    };
    this.handleAdd = this.handleAdd.bind(this);
    this.handleRemove = this.handleRemove.bind(this);
    this.handleOnDragOver = this.handleOnDragOver.bind(this);
    this.handleOnDrop = this.handleOnDrop.bind(this);
  }

  handleAdd = () => {
    const maxIdPerCategory = Math.max(
      ...this.props.category_tasks.map((item: any) => item.id),
      0
    );
    const nextId = maxIdPerCategory + 1;
    const new_task = { id: nextId, text: "Card " + nextId };
    const category_tasks = [new_task, ...this.props.category_tasks];
    this.setState({ category_tasks });
    this.props.createCard(new_task, this.props.category);
  };

  handleRemove = (card_id_to_remove: number) => {
    const category_tasks = this.props.category_tasks.filter((task: any) => {
      return task.id !== card_id_to_remove;
    });
    this.setState({ category_tasks });
    this.props.removeCardFromColumn(card_id_to_remove, this.state.category);
  };

  handleUpdate = (card: any) => {
    const new_task = card;
    const category_tasks = [new_task, ...this.props.category_tasks];
    this.setState({ category_tasks });
    this.props.updateCard(new_task, this.props.category);
  };

  handleOnDragOver(event: any) {
    event.preventDefault();
  }

  handleOnDrop(event: any) {
    let id = event.dataTransfer.getData("id");
    let sourceCategory = event.dataTransfer.getData("sourceCategory");
    this.moveTask(id, sourceCategory);
  }

  moveTask(taskId: number, sourceCategory: string) {
    if (sourceCategory !== this.state.category) {
      this.props.moveCard(taskId, sourceCategory, this.state.category);
    }
  }

  render() {
    const category_tasks = this.props.category_tasks;
    const category = this.props.category ? this.props.category : "";
    const cards = category_tasks.map((task: any, key: number) => {
      return (
        <Card
          key={key}
          id={task.id}
          text={task.text}
          category={category}
          removeCard={this.handleRemove}
          updateCard={this.handleUpdate}
        />
      );
    });

    return (
      <div
        className={"Column " + category}
        onDrop={this.handleOnDrop}
        onDragOver={this.handleOnDragOver}
      >
        <header>
          <h3>{category.split(/(?=[A-Z])/).join(" ")}</h3>
          <button className="add-button" onClick={this.handleAdd}>
            +
          </button>
          <div className="counter">({category_tasks.length})</div>
        </header>
        <div className="task-container droppable">{cards}</div>
      </div>
    );
  }
}
