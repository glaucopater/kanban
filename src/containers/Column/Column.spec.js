import React from "react";
import { shallow } from "enzyme";
import Column from "./Column";

describe("Column", () => {
  const props = { category_tasks: [] };
  let column = shallow(<Column {...props} />);
  it("renders properly", () => {
    expect(column).toMatchSnapshot();
  });
});
