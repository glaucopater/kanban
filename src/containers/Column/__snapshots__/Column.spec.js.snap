// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Column
    category_tasks={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "category": "",
      "children": Array [
        <header>
          <h3>
            
          </h3>
          <button
            className="add-button"
            onClick={[Function]}
          >
            +
          </button>
          <div
            className="counter"
          >
            (
            0
            )
          </div>
        </header>,
        <div
          className="task-container droppable"
        />,
      ],
      "className": "Column ",
      "droppable": "true",
      "onDragOver": [Function],
      "onDrop": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3>
              
            </h3>,
            <button
              className="add-button"
              onClick={[Function]}
            >
              +
            </button>,
            <div
              className="counter"
            >
              (
              0
              )
            </div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
            },
            "ref": null,
            "rendered": "",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "+",
              "className": "add-button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "+",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "(",
                0,
                ")",
              ],
              "className": "counter",
            },
            "ref": null,
            "rendered": Array [
              "(",
              0,
              ")",
            ],
            "type": "div",
          },
        ],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "task-container droppable",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "category": "",
        "children": Array [
          <header>
            <h3>
              
            </h3>
            <button
              className="add-button"
              onClick={[Function]}
            >
              +
            </button>
            <div
              className="counter"
            >
              (
              0
              )
            </div>
          </header>,
          <div
            className="task-container droppable"
          />,
        ],
        "className": "Column ",
        "droppable": "true",
        "onDragOver": [Function],
        "onDrop": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3>
                
              </h3>,
              <button
                className="add-button"
                onClick={[Function]}
              >
                +
              </button>,
              <div
                className="counter"
              >
                (
                0
                )
              </div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "",
              },
              "ref": null,
              "rendered": "",
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "+",
                "className": "add-button",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "+",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "(",
                  0,
                  ")",
                ],
                "className": "counter",
              },
              "ref": null,
              "rendered": Array [
                "(",
                0,
                ")",
              ],
              "type": "div",
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "task-container droppable",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
